### PR TITLE
Workbench dragndrop plotting

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

It is now possible to drag workspaces on to existing figures and have them plot, as MantidPlot currently does. 

**Report to:** [nobody]. 

**To test:**

Using the new workbench, create some workspaces with multiple spectra and others with single spectra. Test:

* plot a line plot, drag another multi-spectrum workspace on to the figure and it should ask which spectra to plot. Entering a selection should overplot that line(s).
* drag a single-spectrum workspace on to the plot and it should plot immediately onto the same figure
* make another figure window and try the same again to ensure the drag events are received by the correct window
* create a colorfill plot of a workspace and drag another workspace onto the same figure - it should replace the existing plot
* try other random things with plotting to see if you can break it

*There is no associated issue.*

*This does not require release notes* because **the workbench will be announced separately.**

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/inc/MantidAPI/AnalysisDataService.h [hunk 1]
```diff
@@ -104,17 +107,21 @@ class DLLExport AnalysisDataServiceImpl final
...
+  enum RetrievalFlags { None = 1 << 0, UnrollGroups = 1 << 2 };
+
   /// Return the list of illegal characters as one string
   const std::string &illegalCharacters() const;
   /// Set the list of illegal characters
   void setIllegalCharacterList(const std::string &);
   /// Is the given name a valid name for an object in the ADS
   const std::string isValid(const std::string &name) const;
-  /// Overridden add member to attach the name to the workspace when a workspace
+  /// Overridden add member to attach the name to the workspace when a
```
Comment (ElliotAOram) [inline Framework/API/inc/MantidAPI/AnalysisDataService.h]: Nit: Are these comment changes from clang? Might be worth neatening these up so they are not spread over more lines than required

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp [hunk 2]
```diff
@@ -1,7 +1,14 @@
+#include "MantidPythonInterface/kernel/Converters/CloneToNumpy.h"
```
Comment (ElliotAOram) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp]: Where is the `CloneToNumpy` used?

**Comment by ElliotAOram** [comment]
Comment on UX (Unrelated to this PR): When you run a algorithm and switch to the Plots tab in the bottom left, you lose the progress bar for the algorithm completion. Is this desired behaviour we want?

**Comment by ElliotAOram** [comment]
*Unhandled exception* (steps to repeat)
```
ws = CreateWorkspace('0,1,2,3', '1,2,3')
empty = CreateEmptyTableWorkspace()
```
* plot `ws`
* drag `empty` on to plot for `ws`

*Side Note: you also get an unhandled exception from plotting an empty table workspace as well*

**Comment by ElliotAOram** [comment]
Unexpected behaviour (probably a separate issue): You can continue to over plot with the same workspace over and over again (adding more and more lines) with the same line name (different colour). I assume this is because the plot has it's own model rather than being linked to the model from where the data comes from (the workspace)?
This in itself is not necessarily an issue however, when you go into the `Figure Options > Curves`, the drop down box only shows 3 of the curves.
image

**Review by ElliotAOram — CHANGES_REQUESTED** [review]
Functionality works with the exception parts of I have commented. Note that some of these comments may well be separate issues not to be addressed in this PR

File: qt/applications/workbench/workbench/plotting/functions.py [hunk 3]
```diff
@@ -103,8 +101,30 @@ def wsname(w):
...
+    :param fig: If not None then use this figure object to plot
+    :return: The figure containing the plot
+    """
+    workspaces = AnalysisDataService.Instance().retrieveWorkspaces(names, unrollGroups=True)
+    selection = get_spectra_selection(workspaces)
+    if selection is not None:
+        return plot(selection.workspaces, spectrum_nums=selection.spectra,
+                    wksp_indices=selection.wksp_indices,
+                    errors=errors, overplot=overplot, fig=fig)
+    else:
```
Comment (ElliotAOram) [inline qt/applications/workbench/workbench/plotting/functions.py]: Nit: redundant `else` statement

**Comment by ElliotAOram** [comment]
Side note on focusing plots. Do we have functionality in the workbench for plots to always remain above the workbench GUI. This can be annoying when using dragging as you refocus on the workbench GUI and hence lose focus on the plot (which if they are overlayed - means it disappears)

**Comment by martyngigg** [comment]
@ElliotAOram Thanks for the thorough testing. I think all of the code comments are valid and I'll address those. As to the other points/questions:

* the disappearing algorithm progress bar sounds like a bug we need to address: #23262
* the overplotting of the same workspace is probably something I should consider here as that behaviour is confusing and we need to decide what we should do.
* similarly with the `TableWorkspace` objects we need to gracefully fail with these too. I'll address this here.

File: Framework/API/inc/MantidAPI/AnalysisDataService.h (hunk 1 above)
Comment (martyngigg) [reply Framework/API/inc/MantidAPI/AnalysisDataService.h]: :+1:

File: Framework/API/inc/MantidAPI/AnalysisDataService.h [hunk 4]
```diff
@@ -104,17 +107,21 @@ class DLLExport AnalysisDataServiceImpl final
   //@}
 
 public:
+  enum RetrievalFlags { None = 1 << 0, UnrollGroups = 1 << 2 };
```
Comment (martyngigg) [inline Framework/API/inc/MantidAPI/AnalysisDataService.h]: I accidentally left this around too...

File: Framework/API/test/AnalysisDataServiceTest.h [hunk 5]
```diff
@@ -75,6 +76,62 @@ class AnalysisDataServiceTest : public CxxTest::TestSuite {
     TS_ASSERT_THROWS(ads.retrieve("z"), Exception::NotFoundError);
   }
 
+  void test_retrieveWorkspaces_with_empty_list_returns_empty_list() {
+    std::vector<Workspace_sptr> empty;
+    TS_ASSERT_EQUALS(empty, ads.retrieveWorkspaces({}));
+    TS_ASSERT_EQUALS(empty, ads.retrieveWorkspaces({}));
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/API/test/AnalysisDataServiceTest.h]: :+1:

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp (hunk 2 above)
Comment (martyngigg) [reply Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataService.cpp]: Good catch. I attempted this a different way at first and then backed out.

File: qt/applications/workbench/workbench/plotting/functions.py (hunk 3 above)
Comment (martyngigg) [reply qt/applications/workbench/workbench/plotting/functions.py]: :+1:

**Comment by martyngigg** [comment]
An interesting observation is that MantidPlot allows you to keep plotting the same curve on top. I can't really see that this is useful it just happens to be what occurs if you don't check the current plot contents. I think the issue with the options dialog is that the curve labels are the same. I think it might be better just to ensure we have unique labels for each of the curves...

**Comment by martyngigg** [comment]
The following script also demonstrates the same problem:

```python
w = CreateSampleWorkspace()
fig = plt.figure()
ax = fig.add_subplot(111, projection='mantid')
plt.plot(w, wkspIndex=0)
plt.plot(w, wkspIndex=0)
plt.legend()
plt.show()
```

**Comment by martyngigg** [comment]
@ElliotAOram This should be ready for testing again now. I think we can address the multiple identical curves separately.

**Comment by ElliotAOram** [comment]
Thanks @martyngigg I'll start building it now

**Review by ElliotAOram — APPROVED** [review]
This works well now and handles varying workspace type (tried with matrix/table/md workspace) overplotting). Once its though the build servers it should be good to go!

**Pull request merged**